### DISCUSSION TITLE
[busybox 1.27.2, Linux 2.6.13.1, target-mipsel_gcc-4.6.4_uClibc-0.9.28] Build of busybox applet "ionice" fails: 'SYS_ioprio_set' undeclared

### DISCUSSION THREAD

**Discussion by dreirund** [Freetz-NG/freetz-ng#1041] [Allgemein]
*Follow-up from [here](https://github.com/Freetz-NG/freetz-ng/discussions/1037#discussioncomment-10614000):*

---

Ahoj,

compiling

* busybox version 1.27.2 with
* toolchain "target-mipsel_gcc-4.6.4_uClibc-0.9.28"
* for Fritz!Box 7113,
* busybox applet "ionice" enabled (`FREETZ_BUSYBOX___V127_IONICE=y`),

fails with `util-linux/ionice.c:34:17: error: 'SYS_ioprio_set' undeclared (first use in this function)`:  
```
util-linux/ionice.c: In function 'ioprio_set':
util-linux/ionice.c:34:17: error: 'SYS_ioprio_set' undeclared (first use in this function)
util-linux/ionice.c:34:17: note: each undeclared identifier is reported only once for each function it appears in
util-linux/ionice.c: In function 'ioprio_get':
util-linux/ionice.c:39:17: error: 'SYS_ioprio_get' undeclared (first use in this function)
util-linux/ionice.c:40:1: warning: control reaches end of non-void function [-Wreturn-type]
util-linux/ionice.c: In function 'ioprio_set':
util-linux/ionice.c:35:1: warning: control reaches end of non-void function [-Wreturn-type]
make[3]: *** [scripts/Makefile.build:197: util-linux/ionice.o] Error 1
make[3]: *** Waiting for unfinished jobs....
coreutils/nproc.c: In function 'nproc_main':
coreutils/nproc.c:36:2: warning: implicit declaration of function 'sched_getaffinity' [-Wimplicit-function-declaration]
make[2]: *** [Makefile:742: util-linux] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [make/busybox/busybox.mk:51: source/target-mipsel_gcc-4.6.4_uClibc-0.9.28/busybox-1.27.2/busybox] Terminated
```

Kernel is 2.6.13.1

[&nearr;&hairsp;Here](http://lists.busybox.net/pipermail/busybox/2009-August/070058.html) it is stated about that error:

> *`SYS_ioprio_set` is a syscall that has been added with 2.6.13.*
> 
> *if you use a newer kernel*
>   *than your toolchain is old, you need to update.*

Kernel version should suffice, so maybe an issue about the toolchain that the build system uses?

[This is the `.config` I use](https://github.com/user-attachments/files/16962570/default.config.txt).

Regards!

**Comment by fda77** [comment]
> Kernel version should suffice

Be sure and "grep" the unpacked kernel sources in "build/" for the missing "SYS_ioprio_set".

Maybe the include path is wrong, or if the symbol does not exist - then search for the kernel patch which added it and backport the patch into make/kernel/patches/

**Reply by dreirund** [reply]
> *Be sure and "grep" the unpacked kernel sources in "build/"*

`build/` is empty in that case where the error occurs.  
But `source/` gives something.

`grep` in the whole `freetz-ng` directory:

`grep -r 'SYS_ioprio_set' *`:  
```
make/pkgs/htop/patches/abandon/010-ioprio_support.patch:+#if defined(SYS_ioprio_set)
make/pkgs/htop/patches/abandon/010-ioprio_support.patch:+# define SYSCALL___ioprio_set SYS_ioprio_set
make/pkgs/htop/patches/abandon/010-ioprio_support.patch:-   syscall(SYS_ioprio_set, IOPRIO_WHO_PROCESS, this->pid, ioprio);
make/pkgs/target-tester/src/target-tester.c:	    int test_SYS = SYS_ioprio_set;
source/target-mipsel_gcc-4.6.4_uClibc-0.9.28/busybox-1.27.2/util-linux/ionice.c:	return syscall(SYS_ioprio_set, which, who, ioprio);
source/target-mipsel_gcc-4.6.4_uClibc-0.9.28/busybox-1.27.2/include/platform.h:# define SYS_ioprio_set __NR_ioprio_set
```

`grep -r 'SYS_ioprio_set' source/kernel/ref-ohio-04.67/linux-2.6.13` gives no result.

> *Maybe the include path is wrong, or if the symbol does not exist - then search for the kernel patch which added it and backport the patch into `make/kernel/patches/`*

That all is beyond my knowledge. I do not know C, CPP, or kernel development.

**Reply by fda77** [reply]
Ops, "build" dir was wrong, "source" has kernel sources was you wrote.

 > That all is beyond my knowledge. I do not know C, CPP, or kernel development.

Me too! So just user you "web browser" and search the kernel sources for the commit which added "SYS_ioprio_set", for kernel 2.6.13 go about 19 years back.
As you wrote it was added in 2.6.13.x, maybe it could be backported in freetz
